### DISCUSSION TITLE
Options for embedding a REPL in a host application

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#101] [Discussions] [answered]
@leppie (or anyone else) I would like to create a REPL in my .NET application. Reviewing the DLR console host infrastructure, it seems not well suited to accommodating this, except perhaps by interacting with a separate IronScheme console process via stdin/out/error. I did observe an IConsole interface that looked like perhaps one way to intercept print statements etc. Perhaps someone has found a simple way to do it. There is also some mention on the wiki about remoting, but what, precisely, is meant by that is unclear, though I have some preconceived notion of what remoting on .NET means that don't seem quite correct. 

Any experience anyone has doing this would be greatly appreciated.

Speaking of DLR host infrastructure I did have some good experience creating my own console host subclass, which allowed me to print an application-specific banner and customize how I load search paths, but obviously this works at the command line in a console application, and what I want is to have a REPL in the GUI app, whether connecting over a network protocol, via stdin/stdout, or running in the same process.

**Comment by bcalco** [comment]
Also, is there a way in Scheme code to append to the `(library-path)`? Or can it only be done from the host via `Builtins.AddIncludePath(...)`?

The motivation is to allow users to use `init.ss` to be able to customize a project in Scheme code.

**Reply by leppie** [reply]
It is a `parameter`. So you can modify it by ie `(library-path "../mypath")` .  Default is basically `(library-path '("." "lib/"))`.

If you search for it in the codebase, there are a few examples.

**Reply by bcalco** [reply]
OIC. Thanks!

**Comment by leppie** [comment, marked as answer]
I would ignore anything DLR in the codebase. It will only make pain :D 

As for remoting, that was removed when .NET Core support was added. 

As for a basic REPL, this is what IronScheme uses:

```
  (define (eval-top-level x)
    (clr-guard [e [e (parameterize ((foreground-color 'red)
                                    (current-output-port (current-error-port)))
                       (display "Unhandled CLR exception during evaluation:\n")
                       (display e)
                       (newline))]]
      (call/cc
        (lambda (k)
          (with-exception-handler
            (lambda (e)
              (let ((serious? (or (serious-condition? e) (system-exception? e) (not (condition? e)))))
                (parameterize ((foreground-color (if serious? 'red 'yellow))
                               (current-output-port (current-error-port)))
                  (when serious?
                    (display "Unhandled exception during evaluation:\n"))
                  (display e)
                  (newline))
                (k)))
            (lambda ()
              (parameterize ([allow-library-redefinition #t])
                (eval x (interaction-environment)))))))))
```

It just comes down to `(eval x (interaction-environment))` which is the same as `Eval` in C# (except that will parse the string first as `x`).

**Reply by bcalco** [reply]
Speaking of `call/cc` ...I was wondering if you could provide a concrete example of a use of continuations that fails in IronScheme, mainly for documentation purposes. Many textbook examples work just fine, but I know (because you have said so, and it is tersely referenced in a few comments) that basically more advanced uses of continuations don't work on .NET. Some more expansive explanation of the boundary would be appreciated so that I can give my users a more detailed heads up.

**Reply by leppie** [reply]
If you use it wrong, this will happen :D 
https://github.com/IronScheme/IronScheme/blob/687cde5d4e463a119e4ba28296f2af42a5c4a9df/IronScheme/IronScheme/Runtime/Control.cs#L98

Basically, when you use `set!` to save it outside the scope to call it later. Notably generators do this. 

Anything in this file should trigger it. https://github.com/IronScheme/IronScheme/blob/master/IronScheme/IronScheme.Console/tests/windtests.sps

Try it on Chez to see what it does.

**Reply by bcalco** [reply]
OK that gives me something concrete to explain the limitation in my documentation. And now I *get* what the nature of the limitation is. Thanks!
